// cspell/dist/index.js
"use strict";
var validator_1 = require("./validator");
exports.validateText = validator_1.validateText;
var Settings_1 = require("./Settings");
exports.mergeSettings = Settings_1.mergeSettings;
exports.readSettings = Settings_1.readSettings;
exports.readSettingsFiles = Settings_1.readSettingsFiles;
exports.defaultSettingsFilename = Settings_1.defaultFileName;
var SpellingDictionary_1 = require("./SpellingDictionary");
exports.createSpellingDictionary = SpellingDictionary_1.createSpellingDictionary;
exports.createSpellingDictionaryRx = SpellingDictionary_1.createSpellingDictionaryRx;
exports.getDictionary = SpellingDictionary_1.getDictionary;
var Settings_2 = require("./Settings");
exports.getDefaultSettings = Settings_2.getDefaultSettings;
var TextDocumentSettings_1 = require("./Settings/TextDocumentSettings");
exports.combineTextAndLanguageSettings = TextDocumentSettings_1.combineTextAndLanguageSettings;
var TextDocumentSettings_2 = require("./Settings/TextDocumentSettings");
exports.constructSettingsForText = TextDocumentSettings_2.combineTextAndLanguageSettings;
const Text = require("./util/text");
exports.Text = Text;
const ExclusionHelper = require("./exclusionHelper");
exports.ExclusionHelper = ExclusionHelper;
var LanguageIds_1 = require("./LanguageIds");
exports.getLanguagesForExt = LanguageIds_1.getLanguagesForExt;
