// cspell/dist/SpellingDictionary/DictionaryLoader.js
"use strict";
const wordListHelper_1 = require("../wordListHelper");
const SpellingDictionary_1 = require("./SpellingDictionary");
const loaders = {
    S: loadSimpleWordList,
    W: loadWordList,
    C: loadCodeWordList,
    T: loadTrie,
};
const dictionaryCache = new Map();
function loadDictionary(uri, options) {
    const { type = 'C' } = options;
    const key = [uri, type].join('|');
    if (!dictionaryCache.has(key)) {
        dictionaryCache.set(key, load(uri, type));
    }
    return dictionaryCache.get(key);
}
exports.loadDictionary = loadDictionary;
function load(uri, type) {
    const regTrieTest = /\.trie\b/i;
    type = regTrieTest.test(uri) ? 'T' : type;
    const loader = loaders[type];
    return loader(uri);
}
function loadSimpleWordList(filename) {
    return SpellingDictionary_1.createSpellingDictionaryRx(wordListHelper_1.loadWordsRx(filename));
}
function loadWordList(filename) {
    return SpellingDictionary_1.createSpellingDictionaryRx(wordListHelper_1.loadWordsRx(filename).flatMap(wordListHelper_1.splitLineIntoWordsRx));
}
function loadCodeWordList(filename) {
    return SpellingDictionary_1.createSpellingDictionaryRx(wordListHelper_1.loadWordsRx(filename).flatMap(wordListHelper_1.splitLineIntoCodeWordsRx));
}
function loadTrie(filename) {
    return SpellingDictionary_1.createSpellingDictionaryTrie(wordListHelper_1.loadWordsRx(filename));
}
